crates/block-producer/src/transaction.rs

```diff
@@ -0,0 +1,211 @@
+            input_notes: tx.input_notes().clone().into(),
+            output_notes,
+            block_ref: tx.block_ref(),
+            expiration_block_num: tx.expiration_block_num(),
+            proof: tx.proof().clone(),
+        }
+    }
+
+    pub fn nullifiers(&self) -> impl Iterator<Item = &Nullifier> {
+        let unauthenticated = self.input_notes.unauthenticated.values().map(|note| &note.nullifier);
+        let witnessed = self.input_notes.witnessed.values().map(|note| &note.nullifier);
+        let proven = self.input_notes.proven.iter().map(|note| &note.0);
+
+        unauthenticated.chain(witnessed).chain(proven)
+    }
+
+    pub fn unauthenticated_notes(&self) -> impl Iterator<Item = &NoteId> {
+        self.input_notes.unauthenticated.keys()
+    }
+
+    pub fn account_update(&self) -> &TxAccountUpdate {
+        &self.account_update
+    }
+
+    pub fn account_id(&self) -> AccountId {
+        self.account_update.account_id()
+    }
+
+    pub fn id(&self) -> TransactionId {
+        self.id
+    }
+
+    pub fn input_notes(&self) -> &InputNotes {
+        &self.input_notes
+    }
+
+    pub fn output_notes(&self) -> &BTreeMap<NoteId, OutputNote> {
+        &self.output_notes
+    }
+
+    pub fn witness_note(
+        &mut self,
+        note_id: NoteId,
+        block_proof: BlockInclusionProof,
+        note_proof: NoteInclusionProof,
+    ) -> Option<UnauthenticatedNote> {
+        let note = self.input_notes.unauthenticated.remove(&note_id)?;
+
+        self.input_notes.witnessed.insert(
+            note_id,
+            WitnessedNote {
+                nullifier: note.nullifier,
+                header: note.header,
+                witness: (block_proof, note_proof),
+            },
+        );
+
+        Some(note)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub struct InputNotes {
+    unauthenticated: BTreeMap<NoteId, UnauthenticatedNote>,
+    witnessed: BTreeMap<NoteId, WitnessedNote>,
+    proven: BTreeSet<ProvenNote>,
+}
+
+impl InputNotes {
+    /// Merges `other` into `self`.
+    ///
+    /// # Errors
+    ///
+    /// Errors if there are duplicate input notes.
+    ///
+    /// Note that this action is __not atomic__.
+    pub fn merge(&mut self, other: Self) -> Result<(), InputNotesError> {
+        for (id, note) in other.unauthenticated {
+            if self.unauthenticated.insert(id, note).is_some() || self.witnessed.contains_key(&id) {
+                return Err(InputNotesError::DuplicateUnauthenticatedNote(id));
+            }
+        }
+
+        for (id, note) in other.witnessed {
+            if self.witnessed.insert(id, note).is_some() || self.unauthenticated.contains_key(&id) {
+                return Err(InputNotesError::DuplicateUnauthenticatedNote(id));
+            }
+        }
+
+        for note in other.proven {
+            if !self.proven.insert(note.clone()) {
+                return Err(InputNotesError::DuplicateProvenNote(note.0));
+            }
+        }
```

Comment:
Should this instead track and compare nullifiers instead? That might make more sense looking at it again.

So an additional `BTreeSet<Nullifier>` property.

Reply:
I've decided to do it this way instead.